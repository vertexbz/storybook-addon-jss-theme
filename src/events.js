// @flow
export const SelectTheme = 'selectTheme';
export const SetThemes = 'setThemes';
