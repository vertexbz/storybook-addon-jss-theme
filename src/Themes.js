// @flow
import * as React from 'react';
import { ButtonStyle, RowStyle, SelectedButtonStyle } from './styles';
import * as event from './events';
import type { ThemeType } from './types';

type ThemesProps = $ReadOnly<{
    channel: any,
    api: any
}>;

type ThemesState = {
    theme: ?ThemeType,
    themes: ThemeType[]
};

export default
class Themes extends React.Component<ThemesProps, ThemesState> {
    state = {
        theme: null,
        themes: []
    };

    constructor(...args: *) {
        super(...args);

        // $FlowIgnore
        this.setTheme = this.setTheme.bind(this);
        // $FlowIgnore
        this.setThemes = this.setThemes.bind(this);
        // $FlowIgnore
        this.onReceiveThemes = this.onReceiveThemes.bind(this);
        // $FlowIgnore
        this.renderThemeButton = this.renderThemeButton.bind(this);
    }

    onReceiveThemes(themes: ThemeType[]) {
        this.setThemes(themes);
        if (themes.length > 0) {
            this.setTheme(themes[0]);
        }
    }

    setTheme(theme: ThemeType) {
        this.setState({ theme });
        this.props.channel.emit(event.SelectTheme, theme);
    }

    setThemes(themes: ThemeType[]) {
        this.setState({ themes });
    }

    componentDidMount() {
        this.props.channel.on(event.SetThemes, this.onReceiveThemes);
    }

    componentWillUnmount() {
        this.props.channel.removeListener(event.SetThemes, this.onReceiveThemes);
    }

    renderThemeButton(theme: ThemeType, i: number): React.Element<'div'> {
        const buttonStyle = theme === this.state.theme ? SelectedButtonStyle : ButtonStyle;
        return <div style={buttonStyle} key={i} onClick={this.setTheme.bind(this, theme)}>{theme.name}</div>;
    }

    render(): ?React.Element<'div'> {
        if (!this.state.themes) {
            return null;
        }

        return (
            <div style={RowStyle}>
                {this.state.themes.map(this.renderThemeButton)}
            </div>
        );
    }
}
