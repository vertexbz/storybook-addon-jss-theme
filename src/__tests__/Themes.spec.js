import { mount } from 'enzyme';
import * as React from 'react';
import Themes from '../Themes';
import EventEmitter from 'events';
import * as event from '../events';

describe('Themes', () => {
    // eslint-disable-next-line max-statements
    it('should work properly', () => {
        const channelSpies = {
            emit: jest.spyOn(EventEmitter.prototype, 'emit'),
            on: jest.spyOn(EventEmitter.prototype, 'on'),
            removeListener: jest.spyOn(EventEmitter.prototype, 'removeListener')
        };

        const channel = new EventEmitter();

        const wrapper = mount(<Themes api={null} channel={channel} />);
        expect(wrapper.render()).toMatchSnapshot();
        expect(channelSpies.on).toBeCalledTimes(1);
        expect(channelSpies.emit).not.toBeCalled();
        expect(channelSpies.removeListener).not.toBeCalled();

        const themes = [
            {
                name: 'theme 1',
                variables: {

                }
            },
            {
                name: 'theme 2',
                variables: {

                }
            }
        ];

        channel.emit(event.SetThemes, themes);
        wrapper.update();
        expect(channelSpies.emit).toBeCalledWith(event.SetThemes, themes);
        expect(channelSpies.emit).toBeCalledWith(event.SelectTheme, themes[0]);
        expect(channelSpies.emit).toBeCalledTimes(2);
        channelSpies.emit.mockClear();

        expect(wrapper.render()).toMatchSnapshot();

        const generatedClassNames = wrapper.children().first().prop('classes');
        const buttons = wrapper.find('.' + generatedClassNames.Button);

        buttons.at(1).simulate('click');
        expect(wrapper.render()).toMatchSnapshot();
        expect(channelSpies.emit).toBeCalledWith(event.SelectTheme, themes[1]);
        expect(channelSpies.emit).toBeCalledTimes(1);
        channelSpies.emit.mockClear();

        buttons.at(0).simulate('click');
        expect(wrapper.render()).toMatchSnapshot();
        expect(channelSpies.emit).toBeCalledWith(event.SelectTheme, themes[0]);
        expect(channelSpies.emit).toBeCalledTimes(1);
        channelSpies.emit.mockClear();

        wrapper.unmount();
        expect(channelSpies.removeListener).toBeCalledTimes(1);
    });
});
