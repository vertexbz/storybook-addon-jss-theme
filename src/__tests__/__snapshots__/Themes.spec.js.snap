// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Themes should work properly 1`] = `null`;

exports[`Themes should work properly 2`] = `
<div
  class="Themes-Row-0-1-1"
>
  <div
    class="Themes-Button-0-1-2 Themes-SelectedButton-0-1-3"
  >
    theme 1
  </div>
  <div
    class="Themes-Button-0-1-2"
  >
    theme 2
  </div>
</div>
`;

exports[`Themes should work properly 3`] = `
<div
  class="Themes-Row-0-1-1"
>
  <div
    class="Themes-Button-0-1-2"
  >
    theme 1
  </div>
  <div
    class="Themes-Button-0-1-2 Themes-SelectedButton-0-1-3"
  >
    theme 2
  </div>
</div>
`;

exports[`Themes should work properly 4`] = `
<div
  class="Themes-Row-0-1-1"
>
  <div
    class="Themes-Button-0-1-2 Themes-SelectedButton-0-1-3"
  >
    theme 1
  </div>
  <div
    class="Themes-Button-0-1-2"
  >
    theme 2
  </div>
</div>
`;
